docs/spec_draft.md

```diff
@@ -1267,6 +1268,70 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `body` function to `inputs` and `init_values` along the `dimensions`
+and produces a `result` tensor.
+
+The `body` function has to be associative, and it produces an
+implementation-defined behavior otherwise. Similarly, if `init_values` do not
+form identity under the `body` function, it produces an implementation-defined
+behavior.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | variadic number of scalars of any supported type |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `body`        | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `inputs`, `init_values` have the same element type.
+  * (C2) All `inputs` have the same shape.
+  * (C3) type(`inputs[k]`) $=$ type(`init_values[k]`) for all `k` $\in$ [0, N).
+  * (C4) size(`inputs`) $=$ size(`init_values`) $=$ size(`result`) $=$ N where
+  N >= 1.
+  * (C5) `init_values` have to form an identity under the `body`.
+  * (C6) 0 $\lt$ `dimensions[d]` $\lt$ rank(`inputs[d]`) for all dimension `d`.
+  * (C7) All dimensions in `dimensions` are unique.
+  * (C8) The full function type of `body` is
+  `(T0,..., TN-1, T0, ..., TN-1) -> (T0, ..., TN-1)`, where the first and second
+  set of input types `T0,..., TN-1` corresponds to the element types of `inputs`
+  and `init_values` respectively. The output types `(T0,..., TN-1)` corresponds
+  to the element types of `results`.
+  * (C9) rank(`results[k]`) = rank(`inputs[k]`) - size(`dimensions`) for all `k`
+  $\in$ [0, N).
+  * (C10) dim(`results[k], d`) $=$ dim(`inputs[k], d`) if `d` $\notin$
+  `dimensions`.
```

Comment:
As it's currently written, C10 is incorrect because the `d` in `dim(inputs[k], d)` actually needs to increase as we move past `d`s in `dimensions`. 

The correct formulation would be pretty laborious to write, so I propose that we replace C9 and C10 with a slightly informal formulation, similarly to what we did in #275 in a similar situation:

```
* (C9) `shape(results[k]) = shape(inputs[k])` except that the dimension sizes
of `inputs[k]` corresponding to `dimensions` are not included.
```

Reply:
This is a neat way to put it, thanks.